Source/CarthageKitTests/Fakes/FakeURLSession.swift

```diff
@@ -0,0 +1,41 @@
+import Foundation
+
+final class FakeDownloadTask: NSURLSessionDownloadTask {
+	var started = false
+	override func resume() {
+		self.started = true
+	}
+
+	var cancelled = false
+	override func cancel() {
+		self.cancelled = true
+	}
+}
+
+final class FakeDataTask: NSURLSessionDataTask {
+	var started = false
+	override func resume() {
+		self.started = true
+	}
+
+	var cancelled = false
+	override func cancel() {
+		self.cancelled = true
+	}
+}
+
+final class FakeURLSession: NSURLSession {
```

Comment:
> Is there a video for that talk?

Sadly, no. :disappointed: 

> So, essentially, you're proposing that we reduce the untested/untestable area of code to as small as possible. I'm a fan of that.

:zap:

> My main concern is having too many pure functions, then it becomes non-obvious how everything fits together. For example, when I want to show someone an example of terrible "clean" code, I show them Xcode.swift: Lots (boo!) of pure (yay!) functions, in addition to this 200-line behemoth, and it's really hard to figure out what's going on because everything is (seemingly?) intermingled. And of course, with the way things are currently written, it is necessarily tested in an integration style.

I definitely agree about the code as it stands today. It really lacks good structure. I _think_ that's the problem: that we have the wrong pure functions, not that we have too many of them. But that's just intuition and could easily be wrong.

And if you think the integration tests as you have them in this PR would be helpful in refactoring and cleaning things up so we can better test, then I will happily merge them! If nothing else, they might be a good starting place for a PR to cleanup the code: add the tests, refactor the code, unit test the code, then remove some or all of the integration tests.